packages/frontend/app/services/channels/vote-verifier.ts

```diff
@@ -0,0 +1,61 @@
+import Service, { inject as service } from '@ember/service';
+import VoteChain from 'emberclear/models/vote-chain';
+import CryptoConnector from '../workers/crypto';
+import WorkersService from '../workers';
+import { generateSortedVote } from './-utils/vote-sorter';
+
+export default class VoteVerifier extends Service {
+  @service workers!: WorkersService;
+  crypto?: CryptoConnector;
+
+  async verify(voteToVerify: VoteChain): Promise<boolean> {
+    this.connectCrypto();
+
+    let voteToVerifyActual: Uint8Array = generateSortedVote(voteToVerify);
+    let voteToVerifyActualHash: Uint8Array = await this.crypto!.hash(voteToVerifyActual);
+
+    let voteToVerifyExpectedHash: Uint8Array = await this.crypto!.openSigned(
+      voteToVerify.signature,
+      voteToVerify.key.publicSigningKey
+    );
+
+    if (!this.checkVoteHashes(voteToVerifyActualHash, voteToVerifyExpectedHash)) {
+      return false;
+    }
+
+    if (voteToVerify.previousVoteChain === undefined) {
+      return true;
+    }
+
```

Comment:
should we check whether someone voted for themselves?

example: i vote yes for myself and Caleb. There's no previous and I sign it and put my key in. Passes algorithm. I can add a test to show.

Reply:
i'll PR a test in.

Reply:
Checking that the contents of a vote are valid will happen outside of this. When we receive a vote, we first check that the vote we are given is a legal vote. Then we will pass it off to the verifier to check for tampering of past votes to get to the current state.

Reply:
So that part will be processed in the vote handler, which will still have the sender's information.